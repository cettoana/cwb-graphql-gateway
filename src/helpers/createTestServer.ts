import { ApolloServer } from 'apollo-server';

import CwbAPI from '../datasources/cwb';
import resolvers from '../resolvers';
import typeDefs from '../schema';

const createTestServer = () => {
  const cwbAPI = new CwbAPI();

  const server = new ApolloServer({
    dataSources: () => ({
      cwbAPI,
    }),
    resolvers,
    typeDefs,
  });

  return server;
};

export default createTestServer;
