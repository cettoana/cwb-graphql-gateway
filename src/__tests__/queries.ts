import { createTestClient } from 'apollo-server-testing';
import dotenv from 'dotenv';
import nock from 'nock';

import createTestServer from '../helpers/createTestServer';
import mockRainfallObsRes from './data/rainfallObsRes.json';
import mockWeatherObsRes from './data/weatherObsRes.json';

dotenv.config();

const queryWeatherObservations = `
query {
  weatherObservations {
    station {
      id
      name
    }
    temperature
  }
}
`;

const queryWeatherObservationWithStationId = `
query {
  weatherObservations(stationId: "CM0030") {
    station {
      id
      name
    }
    temperature
  }
}
`;

const queryWeatherObservationWithStationNames = `
query {
  weatherObservations(stationNames: ["西嶼","東莒"]) {
    station {
      id
      name
    }
    temperature
  }
}
`;

const queryRainfallObservations = `
query {
  rainfallObservations {
    station {
      id
      name
    }
    cumulativeRainfall {
      _10min
      _24hr
    }
  }
}
`;

const queryRainfallObservationsWithStationId = `
query {
  rainfallObservations(stationId: "466940") {
    station {
      id
      name
    }
    cumulativeRainfall {
      _10min
      _24hr
    }
  }
}
`;

const queryRainfallObservationsWithStationNames = `
query {
  rainfallObservations(stationNames: ["板橋","淡水"]) {
    station {
      id
      name
    }
    cumulativeRainfall {
      _10min
      _24hr
    }
  }
}
`;

describe('queries', () => {
  it('query all weather observations', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0001-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
      })
      .reply(200, mockWeatherObsRes);

    const res = await query({ query: queryWeatherObservations });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });

  it('query weather observation with station id', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0001-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
      })
      .reply(200, mockWeatherObsRes);

    const res = await query({ query: queryWeatherObservationWithStationId });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });

  it('query weather observation with location name', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);
    const stationNames = ['西嶼', '東莒'];

    const filteredObsRes = { ...mockWeatherObsRes };
    filteredObsRes.records.location = filteredObsRes.records.location.filter(
      l => stationNames.indexOf(l.locationName) > -1
    );

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0001-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
        locationName: stationNames.join(','),
      })
      .reply(200, filteredObsRes);

    const res = await query({
      query: queryWeatherObservationWithStationNames,
    });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });

  it('query all rainfall observations', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0002-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
      })
      .reply(200, mockRainfallObsRes);

    const res = await query({ query: queryRainfallObservations });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });

  it('query rainfall observation with station id', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0002-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
      })
      .reply(200, mockRainfallObsRes);

    const res = await query({ query: queryRainfallObservationsWithStationId });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });

  it('query rainfall observation with location name', async () => {
    const server = createTestServer();
    const { query } = createTestClient(server);
    const stationNames = ['板橋', '淡水'];

    const filteredObsRes = { ...mockRainfallObsRes };

    filteredObsRes.records.location = mockRainfallObsRes.records.location.filter(
      l => stationNames.indexOf(l.locationName) > -1
    );

    const scope = nock('https://opendata.cwb.gov.tw/api/v1/rest/datastore')
      .get('/O-A0002-001')
      .query({
        Authorization: process.env.CWB_API_KEY,
        format: 'JSON',
        locationName: stationNames.join(','),
      })
      .reply(200, filteredObsRes);

    const res = await query({
      query: queryRainfallObservationsWithStationNames,
    });

    scope.isDone();

    expect(res).toMatchSnapshot();
  });
});
