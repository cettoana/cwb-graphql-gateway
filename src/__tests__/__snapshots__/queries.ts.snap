// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries query all rainfall observations 1`] = `
Object {
  "data": Object {
    "rainfallObservations": Array [
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 24,
        },
        "station": Object {
          "id": "466880",
          "name": "板橋",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 15.5,
        },
        "station": Object {
          "id": "81I520",
          "name": "九份二山",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0.5,
          "_24hr": 19,
        },
        "station": Object {
          "id": "C1I230",
          "name": "九份二山",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 21.5,
        },
        "station": Object {
          "id": "466930",
          "name": "竹子湖",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 23,
        },
        "station": Object {
          "id": "01A420",
          "name": "竹子湖",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 14,
        },
        "station": Object {
          "id": "466900",
          "name": "淡水",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 14.5,
        },
        "station": Object {
          "id": "466940",
          "name": "基隆",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`queries query all weather observations 1`] = `
Object {
  "data": Object {
    "weatherObservations": Array [
      Object {
        "station": Object {
          "id": "C0W120",
          "name": "西嶼",
        },
        "temperature": 16.7,
      },
      Object {
        "station": Object {
          "id": "CM0010",
          "name": "花博豐原",
        },
        "temperature": 13.8,
      },
      Object {
        "station": Object {
          "id": "CM0030",
          "name": "花博外埔",
        },
        "temperature": 12.8,
      },
      Object {
        "station": Object {
          "id": "CM0040",
          "name": "花博后里",
        },
        "temperature": null,
      },
      Object {
        "station": Object {
          "id": "C0W110",
          "name": "東莒",
        },
        "temperature": 8.2,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`queries query rainfall observation with location name 1`] = `
Object {
  "data": Object {
    "rainfallObservations": Array [
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 24,
        },
        "station": Object {
          "id": "466880",
          "name": "板橋",
        },
      },
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 14,
        },
        "station": Object {
          "id": "466900",
          "name": "淡水",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`queries query rainfall observation with station id 1`] = `
Object {
  "data": Object {
    "rainfallObservations": Array [
      Object {
        "cumulativeRainfall": Object {
          "_10min": 0,
          "_24hr": 14.5,
        },
        "station": Object {
          "id": "466940",
          "name": "基隆",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`queries query weather observation with location name 1`] = `
Object {
  "data": Object {
    "weatherObservations": Array [
      Object {
        "station": Object {
          "id": "C0W120",
          "name": "西嶼",
        },
        "temperature": 16.7,
      },
      Object {
        "station": Object {
          "id": "C0W110",
          "name": "東莒",
        },
        "temperature": 8.2,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`queries query weather observation with station id 1`] = `
Object {
  "data": Object {
    "weatherObservations": Array [
      Object {
        "station": Object {
          "id": "CM0030",
          "name": "花博外埔",
        },
        "temperature": 12.8,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
